src/CodeCoverage.php

```diff
@@ -484,6 +486,16 @@ private function applyFilter(RawCodeCoverageData $data): void
         }
     }
 
+    private function applyExecutableLinesFilter(RawCodeCoverageData $data): void
+    {
+        foreach (array_keys($data->lineCoverage()) as $filename) {
+            $data->keepCoverageDataOnlyForLines(
+                $filename,
+                $this->uncoveredFileAnalyser()->executableLinesIn($filename)
```

Comment:
I think there will also a performance impact from calling `executableLinesIn` so much that it might be a good idea to try and mitigate - *if* someone has caching enabled, then that will read from file each time but it might be a good idea to add a memory cache somewhere? Reads on Windows particularly are notoriously slow, and I think this would do an additional read per source file per test if my code-skim is correct?

Reply:
Current
```
    public function enterNode(Node $node): void
    {
        if (!$this->isExecutable($node)) {
            return;
        }

        $this->executableLines[] = $node->getStartLine();
    }

    /**
     * @psalm-return list<int>
     */
    public function executableLines(): array
    {
        $executableLines = array_unique($this->executableLines);

        sort($executableLines);

        return $executableLines;
    }
```

Faster?
```


    public function enterNode(Node $node): void
    {
        if (!$this->isExecutable($node)) {
            return;
        }

        $this->executableLines[$node->getStartLine()] = $node->getStartLine();
    }

    /**
     * @psalm-return list<int>
     */
    public function executableLines(): array
    {
        return $executableLines;
    }
```

Reply:
There's an interesting post by an MS engineer at https://github.com/Microsoft/WSL/issues/873#issuecomment-425272829 in the context of WSL that compares/contrasts performance of the filesystem in Linux vs Windows. Suffice to say it's real

Windows 11, PHP 8.1.2, XDebug 3.1.3, running tests for this repo

9.2 branch
```
Time: 00:19.726, Memory: 360.00 MB

Time: 00:21.713, Memory: 360.00 MB

Time: 00:23.309, Memory: 360.00 MB
```


This PR
```
Time: 00:53.541, Memory: 374.00 MB

Time: 00:44.610, Memory: 374.00 MB

Time: 00:48.638, Memory: 374.00 MB
```

Reply:
Interesting, because we have an in-memory cache and an on-disk cache where the latter is only queried in case of a miss in the former. At least that's how it should be, but [I have been proven wrong](https://thephp.cc/articles/caching-makes-everything-faster-right) about cache related things before.

Reply:
@sebastianbergmann FYI there is an in memory cache for [CachingCoveredFileAnalyser](https://github.com/sebastianbergmann/php-code-coverage/blob/9.2/src/StaticAnalysis/CachingCoveredFileAnalyser.php), but *not* currently for [CachingUncoveredFileAnalyser](https://github.com/sebastianbergmann/php-code-coverage/blob/9.2/src/StaticAnalysis/CachingUncoveredFileAnalyser.php)

That scenario outlined in your blog post of multiple file operations being triggered for every cache call is exactly what's happening here I think.

Reply:
Thanks for the pointer, I'll have a look (hopefully tomorrow morning, otherwise over the weekend).

Reply:
👍

I think the uncovered file analyser not having a memory cache is OK because it's normally only called once at the end of the entire test suite run just before generating the report, unlike the analyser for covered files that's called after every test. It's just that here, the relevant function is now being called from a very different context...

Reply:
Thought about this some more: the idea to separate CoveredFileAnalyzer and UncoveredFileAnalyzer from each other was because the former was required after each test and the latter only once at the end. Now we also need the data from UncoveredFileAnalyzer after each test (right?). If that is really the case, then we should merge the two into FileAnalyzer.

Reply:
I have reverted 3a3285fd68bbd48ad24c229829243d1f0a655a06 (so that this PR no longer has conflicts). I think we should separate caching refactorings/improvements from the topic of this PR.